### ISSUE TITLE
[UI] Binding to app within service creation fails for some services

### ISSUE THREAD

**Issue by thehejik** [epinio/ui#289]
**Describe the bug**
On Epinio v1.9.0 when trying to deploy a new service and binding it to an existing app (in one step) over UI the binding fails. The service itself is deployed.

![image](https://github.com/epinio/epinio/assets/12828077/a779958f-4b2b-435d-911a-bc9d9b0bba45)

```
Service instance was created but failed to reach a deployed state in a reasonable amount of time, no applications were bound
```

@mmartin24 found out that it depends on used service catalog as it works ok with `postgres` and `mongo` but doesn't with `redis`, `mysql` and `rabbitmq`.  It seems that more resource hungry services are unable to bind. Also he was unable to reproduce this issue on Epinio v1.8.1.

I was unable to reproduce that over CLI when trying emulate the UI by:
```
epinio service create mysql-dev mysql-sample --wait && epinio service bind mysql-sample sample
```

**To Reproduce**

1. Install Epinio v1.9.0
2. Push an app over UI
3. Create a mysql-dev service instance and bind it to the app in one step over UI (according to the screenshot)
4. Click `Create` and see error

**Expected behavior**
The service is created and bound to the app

**Comment by richard-cox** [comment]
At the moment we wait for a service `state` of `deployed`. Before this it's only `not ready`

There's a comment in ui code ```Service instance must be `deployed` before they can be bound to apps```.

I tried not waiting at all, having a 2 second gap and also waiting for the resource to be returned via the API, and all give the same 404 response to `/bind` `release 'x443639a38b5fbf3853fe23f3bf7d' does not exist`.

@enrichman Can we add a new `state` to the `service` resource that indicates it's ready to be bound to?

**Comment by mmartin24** [comment]
As @richard-cox  in CLI if no wait is placed in between svc creation and binding, indeed an error appears not finding the the svc:

```
➜  epinio git:(main) epinio service create postgresql-dev postgres-sample && epinio service bind postgres-sample sample

🚢  Creating Service...
Catalog: postgresql-dev
Service: postgres-sample
Wait For Completion: false

🚢  Binding Service...

❌  error binding service: service bind failed: release 'x75ed1dbc13b646923fddfc26d983' does not exist

```
However, please note that in the UI this works for `postgresql` and `mongo` but not for the rest

Now, if we add just a couple of seconds in between this works:
```
➜  epinio git:(main) epinio service create postgresql-dev postgres-sample && sleep 2 && epinio service bind postgres-sample sample

🚢  Creating Service...
Catalog: postgresql-dev
Service: postgres-sample
Wait For Completion: false

🚢  Binding Service...

➜  epinio git:(main) epinio service list                                                                                          

🚢  Listing Services...
Namespace: workspace

✔️  Details:
|      NAME       |            CREATED             | CATALOG SERVICE | VERSION |  STATUS  | APPLICATIONS |
|-----------------|--------------------------------|-----------------|---------|----------|--------------|
| postgres-sample | 2023-08-30 11:52:11 +0200 CEST | postgresql-dev  | 15.1.0  | deployed | sample       |
```

**Comment by enrichman** [comment]
@richard-cox that's strange because as @mmartin24 tested it should work with just a few seconds wait from the async creation. The service resource should be there and we should be able to bind it.

**Comment by richard-cox** [comment]
@enrichman It seems a bit of a workaround to add a wait, which would be susceptible to cluster performance. Would there be a large impact to adding a `created` state? Would the `not ready` state be enough to indicate the resource was created?

**Comment by enrichman** [comment]
It is a workaround. The service should be created as soon the `create` returns, and the `not ready` state should be the status you are looking for (the service is there, but not ready yet). Are you doing the binding in parallel or after the creation?

This should work:
```
service create -> OK -> service bind 
```

if not it's strange. You can avoid the wait with a get, to check for the service:
```
service create -> OK -> while(service get) -> OK -> service bind
```

otherwise the most appropriate way on doing this would be probably have a flag to do this in one API request. This would be an enhancement.

**Comment by richard-cox** [comment]
The current flow is synchronous
- api request to create service
- loop every 2 seconds
  - check if that service state is deployed 
  - fetch service
- when deployed make api request/s to bind app

I tested the above with just waiting for the fetch service to return something. Then also if that state is not-ready. Both still fail.

I've created a PR wait for not-ready....and also to then wait 2 seconds. This isn't great, but resolves the issue

**Comment by thehejik** [comment]
Now it behaves differently but still not OK. 

The service creation with bind usually works fine when the app is freshly deployed but if you delete the service and try to create/bind another service to the same app the action does fail on `configuration not found`. I guess it might be some race condition as the configuration is in some transient state or not created yet. Also I can reproduce this issue on services that were ok before as `postgres`.

![image](https://github.com/epinio/ui/assets/12828077/63647132-0638-4e6e-abc8-9a15a6354c28)

**Comment by richard-cox** [comment]
@thehejik from what i can tell this might be an api issue (https://github.com/epinio/epinio/issues/2551). Separately I spotted edit service is broken and will be fixed by https://github.com/epinio/ui/pull/340

**Comment by mmartin24** [comment]
This looks ok to me.
I tested both issues commented on this ticket and they worked for me:
- the original issue (unable to bind at times when doing so in 1 step on several services) 
- the secondary issue commented [above](https://github.com/epinio/ui/issues/289#issuecomment-1706537496)  not being able to bind after a binded service is deleted.

I checked them first [here](https://github.com/epinio/epinio/issues/2551#issuecomment-1721352133) and rechecked again today with `v1.10.0-rc1-10-g09a409e9` and latest epinio ui, by creating service and binding in 1 step, deleting binded service and repeating the service creation and binding again in 1 steps.

So looks ok in my opinion.

@thehejik , ok with closing this issue?
